src: drm: This commit adds the ability to retrieve and decode EDID binary files.

fonseca54nh opened the pull request.

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
```diff
@@ -1,7 +1,10 @@
 . "${KW_LIB_DIR}/lib/remote.sh" --source-only
 . "${KW_LIB_DIR}/lib/kwlib.sh" --source-only
 
+
```
Do we need this extra new line?

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
```diff
@@ -44,6 +51,22 @@ function drm_main()
   [[ -n "${options_values['VERBOSE']}" ]] && flag='VERBOSE'
 
+  if [[ -n "$decode_edid" ]]; then
+    #module_control 'LOAD' "$target" "$remote" "$load_module" "$flag"
```
Why is this comment here?

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
```diff
@@ -44,6 +51,22 @@ function drm_main()
+    #module_control 'LOAD' "$target" "$remote" "$load_module" "$flag"
+    decode_edid
+    if [[ "$?" != 0 ]]; then
+      return 22
```
Please, add `# EINVAL` to all the returns comments (if you use something different from 22, use the correct label).

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
```diff
@@ -44,6 +51,22 @@ function drm_main()
+    #module_control 'LOAD' "$target" "$remote" "$load_module" "$flag"
+    get_edid
+    if [[ "$?" != 0 ]]; then
+      return 22
```
see my previous comment.

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
```diff
@@ -349,9 +372,65 @@ function get_supported_mode_per_connector()
   printf '%s\n' "$modes"
 }
 
+# Creates a binary EDID file, for each connected monitor, on the current folder.
```
How about something like this:

Capture the EDID file from each connected monitors.

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
```diff
@@ -349,9 +372,65 @@ function get_supported_mode_per_connector()
+{
+  declare -a monitors
+
+  monitors=( $(ls -d /sys/class/drm/*/edid) )
```
Please, check the shellcheck warning for this line. It looks correct to me.

rodrigosiqueira replied:
Avoid the use of `ls` since it can have different outputs in different distros. Use the `find` command. Also, check:

https://kworkflow.org/content/codingstyle.html#avoid-short-options

rodrigosiqueira reviewed (changes requested):
Hi @fonseca54nh ,

Thanks for your PR. Could you address all the shellchecks suggestions? Overall, spellcheck provides good suggestions. Also, it is hard to review all of those messages.

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
In your commit message, reduce the commit title (you don't need 'This commit' for example). Also, add a commit explanation and a reference to the issue if we have one.

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
```diff
@@ -349,9 +372,65 @@ function get_supported_mode_per_connector()
+
+  monitors=( $(ls -d /sys/class/drm/*/edid) )
+
+  say ""
```
I don't think you need this say, just use `\n` or something like that when you print the final result.

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
```diff
@@ -349,9 +372,65 @@ function get_supported_mode_per_connector()
+  say ""
+
+  if [ ${#monitors[@]} -lt 1 ]; then
+	  complain "Could not find any listed monitors!"
```
I think the indentation here is wrong and uses single quotes.

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
```diff
@@ -349,9 +372,65 @@ function get_supported_mode_per_connector()
+  fi
+
+  for i in ${monitors[@]};
+  do
```
This `do` must be in the same line as the `for`.

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
```diff
@@ -349,9 +372,65 @@ function get_supported_mode_per_connector()
+
+  for i in ${monitors[@]};
+  do
+          echo $i | awk -F"/" '{print "Wrote: " $5".bin"}'
```
Please, do not use `echo`, also check:

https://kworkflow.org/content/codingstyle.html#avoid-short-options
https://kworkflow.org/content/codingstyle.html#use-printf-instead-of-echo

rodrigosiqueira commented on src/plugins/subsystems/drm/drm.sh:
```diff
@@ -349,9 +372,65 @@ function get_supported_mode_per_connector()
+		  say $i". "${edids[$i]}
+	  done
+          selection=$(ask_with_default "Decode which file: [0-$((${#edids[@]}-1))]" "0")
+	  edid-decode < ${edids[$selection]}
```
`edid-decode` it is not part of the kw dependencies, you have to add it for the deploy and the local case.

rodrigosiqueira reviewed (changes requested):
Hi @fonseca54nh ,
I added some comments, but first, I recommend that you spend some time trying to address the shellcheck issues and read this page:

https://kworkflow.org/content/codingstyle.html#

fonseca54nh commented:
I've addressed all of your comments regarding shellcheck issues. However, arch linux does not provide the 'edid-decode' package, unlike Debian and Dedora. Instead, it provides the 'read-edid' package, which provides the parse-edid executable. So, even though I'm able to include it as a dependency, I'm unsure on how to call it inside the decode-edid function.

rodrigosiqueira commented:
@fonseca54nh 
Is `parse-edid` available in all distros? If so, we can use it instead of the `edid-decode`.

fonseca54nh commented:
@rodrigosiqueira I've spent some time researching this. I was able to find the package in all 3 distros, with 2 considerations.

1. Fedora names it differently. So instead of using read-edid, on Fedora we should invoke the command as `monitor-parse-edid`.
2. I've found parse-edid's output to be way smaller. Below, I've attached both to be compared. At first glance, I don't know if parse-edid's output is a decent replacement for edid-decode, let me know your thoughts about this.

> edid-decode's output:

```
edid-decode (hex):

00 ff ff ff ff ff ff 00 4c 2d 2d 0d 36 55 32 59
1d 1f 01 04 a5 34 1d 78 22 52 95 a5 56 54 9d 25
0e 50 54 bb 8c 00 b3 00 81 c0 81 00 81 80 95 00
a9 c0 01 01 01 01 02 3a 80 18 71 38 2d 40 58 2c
45 00 09 25 21 00 00 1e 00 00 00 fd 00 38 48 1e
51 11 00 0a 20 20 20 20 20 20 00 00 00 fc 00 43
32 34 46 33 39 30 0a 20 20 20 20 20 00 00 00 ff
00 48 58 35 52 37 30 32 37 33 30 0a 20 20 00 48

----------------

Block 0, Base EDID:
  EDID Structure Version & Revision: 1.4
  Vendor & Product Identification:
    Manufacturer: SAM
    Model: 3373
    Serial Number: 1496470838
    Made in: week 29 of 2021
  Basic Display Parameters & Features:
    Digital display
    Bits per primary color channel: 8
    DisplayPort interface
    Maximum image size: 52 cm x 29 cm
    Gamma: 2.20
    DPMS levels: Off
    Supported color formats: RGB 4:4:4
    First detailed timing includes the native pixel format and preferred refresh rate
  Color Characteristics:
    Red  : 0.6455, 0.3369
    Green: 0.3281, 0.6152
    Blue : 0.1464, 0.0556
    White: 0.3134, 0.3291
  Established Timings I & II:
    IBM     :   720x400    70.081663 Hz   9:5     31.467 kHz     28.320000 MHz
    DMT 0x04:   640x480    59.940476 Hz   4:3     31.469 kHz     25.175000 MHz
    Apple   :   640x480    66.666667 Hz   4:3     35.000 kHz     30.240000 MHz
    DMT 0x05:   640x480    72.808802 Hz   4:3     37.861 kHz     31.500000 MHz
    DMT 0x08:   800x600    56.250000 Hz   4:3     35.156 kHz     36.000000 MHz
    DMT 0x09:   800x600    60.316541 Hz   4:3     37.879 kHz     40.000000 MHz
    DMT 0x0a:   800x600    72.187572 Hz   4:3     48.077 kHz     50.000000 MHz
    DMT 0x10:  1024x768    60.003840 Hz   4:3     48.363 kHz     65.000000 MHz
    DMT 0x11:  1024x768    70.069359 Hz   4:3     56.476 kHz     75.000000 MHz
  Standard Timings:
    DMT 0x3a:  1680x1050   59.954250 Hz  16:10    65.290 kHz    146.250000 MHz
    DMT 0x55:  1280x720    60.000000 Hz  16:9     45.000 kHz     74.250000 MHz
    DMT 0x1c:  1280x800    59.810326 Hz  16:10    49.702 kHz     83.500000 MHz
    DMT 0x23:  1280x1024   60.019740 Hz   5:4     63.981 kHz    108.000000 MHz
    DMT 0x2f:  1440x900    59.887445 Hz  16:10    55.935 kHz    106.500000 MHz
    DMT 0x53:  1600x900    60.000000 Hz  16:9     60.000 kHz    108.000000 MHz (RB)
  Detailed Timing Descriptors:
    DTD 1:  1920x1080   60.000000 Hz  16:9     67.500 kHz    148.500000 MHz (521 mm x 293 mm)
                 Hfront   88 Hsync  44 Hback  148 Hpol P
                 Vfront    4 Vsync   5 Vback   36 Vpol P
    Display Range Limits:
      Monitor ranges (GTF): 56-72 Hz V, 30-81 kHz H, max dotclock 170 MHz
    Display Product Name: 'C24F390'
    Display Product Serial Number: 'HX5R702730'
Checksum: 0x48
```
> Parse-edid's output:

```Section "Monitor"
	Identifier "C24F390"
	ModelName "C24F390"
	VendorName "SAM"
	# Monitor Manufactured week 29 of 2021
	# EDID version 1.4
	# Digital Display
	DisplaySize 520 290
	Gamma 2.20
	Option "DPMS" "true"
	Horizsync 30-81
	VertRefresh 56-72
	# Maximum pixel clock is 170MHz
	#Not giving standard mode: 1680x1050, 60Hz
	#Not giving standard mode: 1280x720, 60Hz
	#Not giving standard mode: 1280x800, 60Hz
	#Not giving standard mode: 1280x1024, 60Hz
	#Not giving standard mode: 1440x900, 60Hz
	#Not giving standard mode: 1600x900, 60Hz
	Modeline 	"Mode 0" +hsync +vsync 
EndSection
```

rodrigosiqueira commented:
Which options are you using for parse-edid? Maybe there are some extra options that show more details?

If only ArchLinux does not have edid-decoded, maybe we can add a specific implementation only for ArchLinux that will use parse-edid. We can also add a check in the ArchLinux case where we check if the user has edit-decoded in their system base; in that case, we just use it.

davidbtadokoro commented:
Hi @fonseca54nh, any updates here?

fonseca54nh commented:
Sorry for the delay. Unfortunately, parse-edid has no extra options.

rodrigosiqueira commented:
@fonseca54nh could you rebase this change against the latest unstable branch?

Also, parse-edid is the common denominator, right? If so, just use it.

rodrigosiqueira commented:
Closed due to the lack of activity.
